[docs] Creates accessibility page with content for date pickers (#13274)
mui/mui-x: docs/data/date-pickers/accessibility/accessibility.md

| <kbd class="key">Arrow Up</kbd>, <kbd class="key">Arrow Down</kbd>,<br> <kbd class="key">Arrow Left</kbd>, <kbd class="key">Arrow Right</kbd> | Moves focus among the available values                        |
|                                                                                                                <kbd class="key">Page Up</kbd> | Moves focus to the last day of the month                      |
|                                                                                                              <kbd class="key">Page Down</kbd> | Moves focus to the first day of the month                     |
|                                                                                                                   <kbd class="key">Home</kbd> | Moves focus to first day of the week within the current month |
|                                                                                                                    <kbd class="key">End</kbd> | Moves focus to last day of the week within the current month  |

### Date Range Picker

> When interacting with the keyboard, the [Date Range Picker](/x/react-date-pickers/date-range-picker/) keep the focus on the Field component, thereby offering the same keyboard navigation support as the [Date Range Field](/x/react-date-pickers/accessibility/#fields).

Reviewer: ```suggestion
When interacting with the keyboard, the [Date Range Picker](/x/react-date-pickers/date-range-picker/) keeps the focus on the Field component, thereby offering the same keyboard navigation support as the [Date Range Field](/x/react-date-pickers/accessibility/#fields).
```
P.S. WDYT about removing the `Date Range Field` or the `Field` link below on L84? 🤔 
They point to the same section and are too close to one another IMHO. 🙈

Author: I agree, and actually I think L82 and L84 are redundant

---
productId: x-date-pickers
title: Date and Time Pickers - Accessibility
githubLabel: 'component: pickers'
packageName: '@mui/x-date-pickers'
---

# Accessibility

<p class="description">The Date and Time Pickers have complete accessibility support, including built-in keyboard navigation that follows international standards.</p>

## Guidelines

> The most commonly encountered conformance guidelines for accessibility are:

Reviewer: Nitpick: This section is a copy and paste from Data Grid and Tree View a11y pages.
If we apply this change, it would be nice to replicate it on those pages as well. 👍

> The most commonly encountered conformance guidelines for accessibility are:

Reviewer: @arthurbalduini do you plan on applying these changes?
Or are you leaving that for a separate PR?

> The most commonly encountered conformance guidelines for accessibility are:

Author: After discussing with @samuelsycamore, I will apply these changes and also replicate them to the other pages where they appear on this PR

WCAG 2.1 has three levels of conformance: A, AA, and AAA.
Level AA meets the most commonly encountered conformance guidelines.
This is the most common target for organizations so it is what we aim to support.

The WAI-ARIA Authoring Practices includes examples on [Date Picker Dialog](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/datepicker-dialog/) and [Date Picker Spin Button](https://www.w3.org/WAI/ARIA/apg/patterns/spinbutton/examples/datepicker-spinbuttons/) that provide valuable information on how to optimize the accessibility of the Date and Time Pickers.

## Screen reader compatibility

Date and Time Pickers use ARIA roles and robust focus management across the interactive elements to convey the necessary information to users, being optimized for use with assistive technologies.

## Keyboard support

The Date and Time Pickers consist of different associations of Fields, Calendar, and Clock components. Each of these components is designed to respond intuitively to keyboard interactions, providing extensive keyboard navigation support.

The following sections explain in detail the keyboard support for various components:

### Fields

> The following table describes the keyboard support for all [field components](/x/react-date-pickers/fields/)

Reviewer: ```suggestion
The following table describes the keyboard support for all [field components](/x/react-date-pickers/fields/):
```

|                                                                                       <kbd class="key">Shift</kbd>+<kbd class="key">Tab</kbd> | Moves focus to the previous element    |
| <kbd class="key">Arrow Up</kbd>, <kbd class="key">Arrow Down</kbd>,<br> <kbd class="key">Arrow Left</kbd>, <kbd class="key">Arrow Right</kbd> | Moves focus among the available values |

In addition to the basic keyboard support above, the day view component implements its own keyboard interactions:

|                             Keys | Description                                                       |
| -------------------------------: | :---------------------------------------------------------------- |
|   <kbd class="key">Page Up</kbd> | Moves calendar to next month, keeping focus on the same day       |
| <kbd class="key">Page Down</kbd> | Moves calendar to previous month, keeping focus on the same day   |
|      <kbd class="key">Home</kbd> | Moves focus to the first day of the week within the current month |

> |       <kbd class="key">End</kbd> | Moves focus to the last day of the week within the current month  |

Reviewer: Are you sure about the `current month` part here?
I'm pretty sure that it applies to `DateRangeCalendar` instead. 🙈

| --------------------------------------------------------------------------------------------------------------------------------------------: | :---------------------------------------- |
| <kbd class="key">Arrow Up</kbd>, <kbd class="key">Arrow Down</kbd>,<br> <kbd class="key">Arrow Left</kbd>, <kbd class="key">Arrow Right</kbd> | Moves focus among the available values    |
|                                                                                                                <kbd class="key">Page Up</kbd> | Moves focus to the last day of the month  |
|                                                                                                              <kbd class="key">Page Down</kbd> | Moves focus to the first day of the month |
|                                                                                                                   <kbd class="key">Home</kbd> | Moves focus to first day of the week      |
|                                                                                                                    <kbd class="key">End</kbd> | Moves focus to last day of the week       |

### Date Range Picker

> The [Date Range Picker](/x/react-date-pickers/date-range-picker/) keeps the focus on the field all the time and thus has the same keyboard navigation as Date Range [Field](/x/react-date-pickers/accessibility/#fields).

Reviewer: This references the current Date Range Picker UX pattern, where the focus is returned to the input after selection in an open Picker view as compared to Date Picker, where the focus is kept in the Popper until the complete selection is made.

> The [Date Range Picker](/x/react-date-pickers/date-range-picker/) keeps the focus on the field all the time and thus has the same keyboard navigation as Date Range [Field](/x/react-date-pickers/accessibility/#fields).

Author: I have rewritten this part on [#c10ee9f](https://github.com/mui/mui-x/pull/13274/commits/c10ee9f5d0457fc743859d001c591fb5ea86dffc). Would appreciate your opinion ! :)

---
productId: x-date-pickers
title: Date and Time Pickers - Date format and localization
githubLabel: 'component: pickers'
packageName: '@mui/x-date-pickers'
---

# Date and Time Pickers - Accessibility

<p class="description">The Date and Time Pickers have complete accessibility support, including built-in keyboard navigation that follows international standards.</p>

## Guidelines

The most commonly encountered conformance guidelines for accessibility are:

- Globally accepted standard: [WCAG](https://www.w3.org/WAI/standards-guidelines/wcag/)
- US:
  - [ADA](https://www.ada.gov/) - US Department of Justice
  - [Section 508](https://www.section508.gov/) - US federal agencies
- Europe: [EAA](https://ec.europa.eu/social/main.jsp?catId=1202) (European Accessibility Act)

WCAG 2.1 has three levels of conformance: A, AA, and AAA.
Level AA meets the most commonly encountered conformance guidelines.

> This is the most common target for organizations so what we aims to support very well.

Reviewer 2: ```suggestion
This is the most common target for organizations so it is what we aim to support very well.
```
"Very well" still seems a bit off. 🤔

Author: Good point ! 👍

<!-- Those examples do not cover the clock component -->

> 

Author: Agreed ! 👍

The WAI-ARIA Authoring Practices includes examples on [Date Picker Dialog](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/datepicker-dialog/) and [Date Picker Spin Button](https://www.w3.org/WAI/ARIA/apg/patterns/spinbutton/examples/datepicker-spinbuttons/) that provide valuable information on how to optimize the accessibility of the Date and Time Pickers.

## Screen reader compatibility

Date and Time Pickers provide support to ARIA roles and a robust focus management across the interactive elements to convey the necessary information to users, being optimized for use with assistive technologies.

## Keyboard support

> The Date and Time Pickers consist on different associations of Fields, Calendar, Time and Digital Clock components, each of them designed to respond intuitively to keyboard interactions, providing a better experience for users who rely on keyboard navigation.

Reviewer 2: "consist of"? Maybe?

Reviewer 2: The sentence is very long. Can we split it up into at least 2 sentences?

Author: Done in  [8e2b7e3](https://github.com/mui/mui-x/pull/13274/commits/8e2b7e3de0c9884702a21f75b5b00c1dc4c51a5f)

> The Date and Time Pickers consist of different associations of Fields, Calendar, Time and Digital Clock components. Each of these are designed to respond intuitively to keyboard interactions, providing a better experience for users who rely on keyboard navigation.

Author: Nice suggestion !

Each available component presents its own keyboard support, detailed on the following section:

### Fields

The following table describes the keyboard support for all [field components](/x/react-date-pickers/fields/)

|                               Keys | Description                                  |
| ---------------------------------: | :------------------------------------------- |
|  <kbd class="key">Arrow Left</kbd> | Moves focus among date/time sections         |

> | <kbd class="key">Arrow Right</kbd> | Moves focus among date/time sections         |

Author: I agree ! 👍 
And I also think this way to describe the multiple arrow keys is better than the one I did with the parenthesis !

> | <kbd class="key">Arrow Right</kbd> | Moves focus among date/time sections         |

Author: Done in [9d6e753](https://github.com/mui/mui-x/pull/13274/commits/9d6e753a7e53da7cea2fc19e5f77f232de4fddcc)

|    <kbd class="key">Arrow Up</kbd> | Increases focused section value by 1         |
|  <kbd class="key">Arrow Down</kbd> | Decreases focused section value section by 1 |
|     <kbd class="key">Page Up</kbd> | Increases focused section value section by 5 |
|   <kbd class="key">Page Down</kbd> | Decreases focused section value section by 5 |
|        <kbd class="key">Home</kbd> | Sets focused section to the first value      |

> |         <kbd class="key">End</kbd> | Sets focused section to the last value       |

Author: Nice suggestion ! 💙

### Date Calendar

All the available [views](/x/react-date-pickers/date-calendar/#views) implement the same basic keyboard support:

|                                                      Keys | Description                            |
| --------------------------------------------------------: | :------------------------------------- |
|                                <kbd class="key">Tab</kbd> | Moves focus to the next element        |

> |   <kbd class="key">Shift</kbd>+<kbd class="key">Tab</kbd> | Moves focus to the previous element    |

Author: Yes, I put that on purpose to leave a breach to mention the component as a whole before talking about day view's specific keyboard support. 
I agree this part is quite irrelevant. Do you think that on the Date Calendar section we could only talk about the day view ?

> |   <kbd class="key">Shift</kbd>+<kbd class="key">Tab</kbd> | Moves focus to the previous element    |

Reviewer: > Do you think that on the Date Calendar section we could only talk about the day view ?

That is an option, but if you think that mentioning other views brings value, I'm all for it. 👍

> |   <kbd class="key">Shift</kbd>+<kbd class="key">Tab</kbd> | Moves focus to the previous element    |

Author: I refactor the sentence on [#206e723](https://github.com/mui/mui-x/pull/13274/commits/206e723c5524dce7b6cbc652e0018d444d60d3d0) to remove the unnecessary mention of the regular keys support. Your opinion would be extremely valuable to me ! 🙂

| <kbd class="key">Arrow Keys (Left, Right, Up, Down)</kbd> | Moves focus among the available values |

> In addition to the basic keyboard support above, the day view component implements its own keyboard interactions:

Author: I agree, but I also felt that the keyboard support differences among the views was a relevant information 🤔

> In addition to the basic keyboard support above, the day view component implements its own keyboard interactions:

Reviewer: Yes, that is very relevant information, it's just that `day view` sounds and somewhat is like an internal notion, but I'm fine with going for it now and adjusting if necessary or when we think of something better. 👍

> In addition to the basic keyboard support above, the day view component implements its own keyboard interactions:

Author: I assumed that since we mention it [here](https://mui.com/x/react-date-pickers/date-calendar/#views) it was a known therm for the docs
I'm also not so happy with this part, but I couldn't come with a better reference 😢

|                             Keys | Description                                                                       |
| -------------------------------: | :-------------------------------------------------------------------------------- |
|   <kbd class="key">Page Up</kbd> | Moves calendar to next month, keeping focus on the same day                       |
| <kbd class="key">Page Down</kbd> | Moves calendar to previous month, keeping focus on the same day                   |
|      <kbd class="key">Home</kbd> | Moves focus to first day of the week, tresspassing to previous month if necessary |

> |       <kbd class="key">End</kbd> | Moves focus to last day of the week, tresspassing to next month if necessary      |

Reviewer: That's why I made the following suggestions for DateRangeCalendar: https://github.com/mui/mui-x/pull/13274/files/8e2b7e3de0c9884702a21f75b5b00c1dc4c51a5f#r1619236475

It seems that the `DateRangeCalendar` is an exception here, not the `DateCalendar`, that's why I was looking on how to "translate" that piece of information the best. 🤔

### Date Picker

The [Date Picker](/x/react-date-pickers/date-picker/) combines the functionalities of the Date Field and Date Calendar components.

> Depending on which component is in focus, the Picker will provide the corresponding keyboard support, either from [DateField](/x/react-date-pickers/accessibility/#fields) or [DateCalendar](/x/react-date-pickers/accessibility/#date-calendar) and any changes made to the value will be consistently reflected across both components.

Reviewer: WDYT about omitting that extra part of the information? I don't think that it is related to a11y or keyboard navigation. 🤔 
```suggestion
Depending on which component is in focus, the Picker will provide the corresponding keyboard support, either from [DateField](/x/react-date-pickers/accessibility/#fields) or [DateCalendar](/x/react-date-pickers/accessibility/#date-calendar).
```

Author: You are absolutely right !

### Date Range Calendar

> The Date Range Calendar implements a similar keyboard support as day view of the [DateCalendar](/x/react-date-pickers/accessibility/#date-calendar) component, with a difference on the navigation among the previous and next months that must be achieved using the auxiliar arrows.

Reviewer 2: ```suggestion
The Date Range Calendar implements a similar keyboard support as the day view of the [DateCalendar](/x/react-date-pickers/accessibility/#date-calendar) component, with a difference on the navigation among the previous and next months that must be achieved using the auxiliar arrows.
```

> The Date Range Calendar implements a similar keyboard support as the day view of the [DateCalendar](/x/react-date-pickers/accessibility/#date-calendar) component, with a difference on the navigation among the previous and next months that must be achieved using the auxiliar arrows.

Reviewer: ```suggestion
The [Date Range Calendar](/x/react-date-pickers/date-range-calendar/) implements a similar keyboard support as the day view of the [DateCalendar](/x/react-date-pickers/accessibility/#date-calendar) component, with a difference on the navigation among the previous and next months that must be achieved using the arrows in the calendar header.
```

Author: Nice one !

|                                                      Keys | Description                               |
| --------------------------------------------------------: | :---------------------------------------- |
| <kbd class="key">Arrow Keys (Left, Right, Up, Down)</kbd> | Moves focus among the available values    |
|                            <kbd class="key">Page Up</kbd> | Moves focus to the last day of the month  |
|                          <kbd class="key">Page Down</kbd> | Moves focus to the first day of the month |

> |                               <kbd class="key">Home</kbd> | Moves focus to first day of the week      |

Reviewer: ```suggestion
|                               <kbd class="key">Home</kbd> | Moves focus to the first day of the week within the current month     |
```

> |                                <kbd class="key">End</kbd> | Moves focus to last day of the week       |

Reviewer: ```suggestion
|                                <kbd class="key">End</kbd> | Moves focus to the last day of the week within the current month      |
```

### Date Range Picker

> The [Date Range Picker](/x/react-date-pickers/date-range-picker/) integrates the features of both the Date Range Field and Date Range Calendar components. However it doesn't inherently provides every keyboard support implemented in these components.

Reviewer: WDYT about not talking about the calendar here, because it's keyboard interaction is never active in a range picker. 🤔 
```suggestion
The [Date Range Picker](/x/react-date-pickers/date-range-picker/) keeps the focus on the field all the time and thus has the same keyboard navigation as Date Range [Field](/x/react-date-pickers/accessibility/#fields).
```

Author: You suggestion is much clearer ! 🎉

> To ensure a more fluid and more accessible user experience, Date Range Picker support the same keyboard navigation and interactions of the [Field](/x/react-date-pickers/accessibility/#fields) components, having the changes being consistently displayed on the calendar component.

Reviewer 2: ```suggestion
To ensure a more fluid and accessible user experience, Date Range Picker supports the same keyboard navigation and interactions of the [Field](/x/react-date-pickers/accessibility/#fields) components, having the changes being consistently displayed on the calendar component.
```

Author: Thanks, Michel !

> To ensure a more fluid and accessible user experience, Date Range Picker supports the same keyboard navigation and interactions of the [Field](/x/react-date-pickers/accessibility/#fields) components, having the changes being consistently displayed on the calendar component.

Reviewer: ```suggestion
To ensure a more fluid and accessible user experience, Date Range Picker supports the same keyboard navigation and interactions of the [Field](/x/react-date-pickers/accessibility/#fields) components, having the changes consistently updated on the calendar component.
```
